overlord/snapstate/check_snap.go

```diff
@@ -490,17 +601,29 @@ func checkSystemUsernames(si *snap.Info) error {
 		return err
 	}
 
+	extrausers := !release.OnClassic
 	for _, user := range si.SystemUsernames {
-		if !supportedSystemUsernames[user.Name] {
+		var id uint32
+		id, ok := supportedSystemUsernames[user.Name]
+		if !ok {
 			return fmt.Errorf(`snap %q requires unsupported system username "%s"`, si.InstanceName(), user.Name)
 		}
 
 		switch user.Scope {
 		case "shared":
-			_, uidErr := findUid(user.Name)
-			_, gidErr := findGid(user.Name)
-			if uidErr != nil || gidErr != nil {
-				return fmt.Errorf(`snap %q requires that both the "%s" system user and group are present on the system.`, si.InstanceName(), user.Name)
+			// Create the snapd-range-<base>-root user and group so
+			// systemd-nspawn can avoid our range. Our ranges will always
+			// be in 65536 chunks, so mask off the lower bits to obtain our
+			// base (see above)
+			rangeStart := id & 0xFFFF0000
+			rangeName := fmt.Sprintf("snapd-range-%d-root", rangeStart)
+			if err := ensureUserGroup(rangeName, rangeStart, extrausers); err != nil {
+				return fmt.Errorf(`cannot ensure users for snap %q required system username "%s": %v`, si.InstanceName(), user.Name, err)
+			}
+
+			// Create the requested user and group
+			if err := ensureUserGroup(user.Name, id, extrausers); err != nil {
```

Comment:
AFAICT we don't have positive tests that check if the values of "id" and "extrausers" are correctly passed. The user.Name is tested indirectly via the error string which is probably fine.

Reply:
These tests were moved to osutil/user_test.go when moving the lowlevel logic in 88abe5412416245678bc962e20b72574e9b0de0b

Reply:
This will also be covered in a new check I added, TestCheckSnapSystemUsernamesCalls().

Reply:
Done